controller-server/src/main/java/com/yahoo/vespa/hosted/controller/deployment/InternalStepRunner.java

```diff
@@ -566,6 +566,7 @@ private void sendNotification(Run run, DualLogger logger) {
 
     /** Returns the real application with the given id. */
     private Application application(ApplicationId id) {
+        controller.applications().lockOrThrow(id, __ -> { }); // Memory fence.
```

Comment:
Isn't this still a race? Another thread can still update application/deployments after lock is released here, before `require` completes, but maybe not an issue in practice.

Reply:
Correct about everything :)